search/engine.go

```diff
@@ -320,19 +319,152 @@ func (e *Engine) updateConfig(databaseName string, configData io.Reader) error {
+	tableRow := e.IndexDB.QueryRow("SELECT tablename, config FROM search_config WHERE longlived_name = $1", table)
+
+	err := tableRow.Scan(&currentTableName, &tableConfig)
+	if err != nil {
+		e.ErrorLog.Errorf("Unable to get config for table %v", table)
+		return "", nil
+	}
+
+	err = json.Unmarshal([]byte(tableConfig), &configTable)
+	if err != nil {
+		e.ErrorLog.Errorf("Unable to get config for table %v", table)
+		return "", nil
+	}
+
+	return currentTableName, configTable
+}
+
+/*
+	getSrcTableColumns - returns a list of column names from the source table
+*/
+func (e *Engine) getSrcTableColumns(table string) ([]string, error) {
+	mainConfig := e.GetConfig()
+
+	dbConfig, haveDatabase := mainConfig.Databases[genericDatabaseName]
+	if !haveDatabase {
+		return nil, errors.New("No 'generic' database config found")
+	}
+
+	genericDB, err := sql.Open(dbConfig.Driver, dbConfig.DSN())
+	if err != nil {
+		return nil, err
+	}
+	defer genericDB.Close()
+
+	rows, err := genericDB.Query(fmt.Sprintf(`SELECT *FROM %v`, table))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	return rows.Columns()
+}
+
+/*
+	updateConfigTable - updates config tablename using longLivedName
+*/
+func (e *Engine) updateConfigTable(tableName, longLivedName string) error {
+	// If table is not configured for search just return
+	// Nothing to update or check
+	currentTableName, currentTable := e.getConfigTable(longLivedName)
+	if currentTableName == "" && currentTable == nil {
+		return nil
+	}
+
+	columnNames, err := e.getSrcTableColumns(tableName)
+	if err != nil {
+		e.ErrorLog.Errorf("updateConfigTable: unable to get source column names for table %v, %v", tableName, err)
+		return err
+	}
+
+	sort.Strings(columnNames) // Order is important for Binary Search
+
+	for i, field := range currentTable.Fields {
+		// sort.SearchStrings uses Binary Search for strings
+		// For more details https://golang.org/pkg/sort/#SearchStrings
+		index := sort.SearchStrings(columnNames, field.Field)
+		if !(index < len(columnNames) && columnNames[index] == field.Field) {
+			copy(currentTable.Fields[i:], currentTable.Fields[i+1:])
+
+			currentTable.Fields[len(currentTable.Fields)-1] = nil
+			currentTable.Fields = currentTable.Fields[:len(currentTable.Fields)-1]
+		}
+	}
+
+	if len(currentTable.Fields) > 0 {
+		jsonConfig, err := json.Marshal(currentTable)
+		if err != nil {
+			e.ErrorLog.Errorf("updateConfigTable: unable to marshal into json, table %v, %v", tableName, err)
+			return err
+		}
+
+		_, err = e.IndexDB.Exec(`UPDATE search_config SET tablename = $1, config = $2 WHERE longlived_name = $3;`, tableName, jsonConfig, longLivedName)
+		if err != nil {
+			e.ErrorLog.Errorf("updateConfigTable: unable to update search_config for table %v, %v", tableName, err)
+			return err
+		}
+
+		e.ErrorLog.Infof("Config updated for table %v", longLivedName)
+	} else {
+		_, err := e.IndexDB.Exec(`DELETE FROM search_config WHERE longlived_name = $1`, longLivedName)
```

Comment:
Change the log message to be clearer, explaining that the table has to be removed, because none of the originally indexed fields exist anymore.